File: Slide_Window/2653. Sliding Subarray Beauty.cpp
// Sliding Window TC:O(n * U) SC:O(U)
class Solution {
public:
    vector<int> getSubarrayBeauty(vector<int>& nums, int k, int x) {
        const int BASE = 50;
        vector<int> arr(BASE*2+1);
        int cnt = 0;
        vector<int> ans;
        int n = nums.size();
        int i = 0;
        for(int j = 0; j < n; j++) {
            arr[nums[j] + BASE] += 1;
            cnt += (nums[j] < 0);
            if(j - i + 1 == k) {
                if(cnt < x) {
                    ans.push_back(0);
                }
                else {
                    int tmp = 0;
                    for(int i = 0; i < BASE; i++) {
                        tmp += arr[i];
                        if(tmp >= x) {
                            ans.push_back(i - BASE);
                            break;
                        }
                    }
                }
                arr[nums[i] + BASE] -= 1;
                cnt -= (nums[i] < 0);
                i++;
            }
        }

        return ans;
    }
};

// Sliding Window + multiset TC:O(nlgn) SC:O(n)
class Solution {
public:
    vector<int> getSubarrayBeauty(vector<int>& nums, int k, int x) {
        multiset<int> set1;
        multiset<int> set2;
        vector<int> ans;
        int n = nums.size();
        int i = 0;
        for(int j = 0; j < n; j++) {
            if(set1.size() < x) set1.insert(nums[j]);
            else {
                if(nums[j] < *set1.rbegin()) {
                    int tmp = *set1.rbegin();
                    set1.erase(--set1.end());
                    set1.insert(nums[j]);
                    set2.insert(tmp);
                }
                else set2.insert(nums[j]);
            }
            if(j - i + 1 == k) {
                int tmp = *set1.rbegin();
                if(tmp < 0) ans.push_back(tmp);
                else ans.push_back(0);
                if(set1.find(nums[i]) != set1.end()) {
                    set1.erase(set1.lower_bound(nums[i]));
                    if(!set2.empty()) {
                        int tmp1 = *set2.begin();
                        set1.insert(tmp1);
                        set2.erase(set2.begin());
                    }
                }
                else {
                    set2.erase(set2.lower_bound(nums[i]));
                }
                i++;
            }
        }

        return ans;
    }
};

File: Data_Structure/2653. Sliding Subarray Beauty.py
# SortedList or multiset: TC:O(nlgn) SC:O(n) Array: TC:O(50 * n) SC:O(100)  BIT: TC:O(nlgnlgn) SC:O(n)
class FenwickTree:
    def __init__(self, n: int):
        self.tree = [0] * (n + 1) 
    def update(self, i: int, val: int) -> None:
        while i < len(self.tree):
            self.tree[i] += val
            i += i & -i
    def pre(self, i: int) -> int:
        res = 0
        while i > 0:
            res += self.tree[i]
            i &= i - 1
        return res
    def query(self, l: int, r: int) -> int:
        if r < l:
            return 0
        return self.pre(r) - self.pre(l - 1)
class Solution:
    def getSubarrayBeauty(self, nums: List[int], k: int, x: int) -> List[int]:
        '''
        去找每個定長的subarray 第 x 小的負數 如果少於x個neg就是0
        1. sliding window 並且需要維護最小值 會想到啥? 
        a. set C++
        b. SortedList 直接去二分 python
        c. array 從值域下手 去判斷是否有超過x個neg
        2. 一樣做 sliding window 但是用樹狀數組去存 存完對樹狀數組做二分判斷第X小的數字是啥 
        '''
        n = len(nums)
        # BIT 需離散化 一樣數字直接去重 重新定義下標即可
        mp = {x : i + 1 for i, x in enumerate(sorted(set(nums)))}
        arr = sorted(mp.keys())
        bit = FenwickTree(n)
        for num in nums[:k-1]:
            bit.update(mp[num], 1)
        def xth(x : int) -> int:
            l, r = 1, n # x th 小的數值在哪個1-index上
            while l < r:
                mid = (r + l) // 2
                if bit.pre(mid) >= x:
                    r = mid
                else:
                    l = mid + 1
            return min(0, arr[l-1])

        ans = [0] * (n - k + 1)
        for i in range(k-1, n):
            bit.update(mp[nums[i]], 1) # 進入窗口
            # 二分
            ans[i - k + 1] = xth(x)
            bit.update(mp[nums[i - k + 1]], -1) # 出窗
        return ans

        '''
        # array值域 做法
        cnt = [0] * 101
        ans = [0] * (n - k + 1)
        for num in nums[:k - 1]:
            cnt[num] += 1
        for i, (in_, out) in enumerate(zip(nums[k - 1:], nums)):
            cnt[in_] += 1
            left = x
            for j in range(-50, 0):
                left -= cnt[j]
                if left <= 0:
                    ans[i] = j
                    break
            cnt[out] -= 1
        return ans
        '''
            
        '''
        # SortedList
        sl = SortedList()
        i = 0
        ans = []
        for j in range(n):
            sl.add(nums[j])
            if j >= k - 1:
                sm_x = sl[x - 1]
                if sm_x < 0:
                    ans.append(sm_x)
                else:
                    ans.append(0)
            if j - i + 1 == k:
                sl.remove(nums[i])
                i += 1
        return ans
        '''
